Eliminate dependency on deprecated setuptools.pkg_resources

ferdnyc opened the pull request:
This PR eliminates use of setuptools' [deprecated](https://setuptools.pypa.io/en/latest/pkg_resources.html#package-discovery-and-resource-access-using-pkg-resources) `pkg_resources` in `hotdoc.utils.utils` and `hotdoc.extensions.gi.utils`, substituting equivalent APIs provided by `importlib.metadata.entry_points()`.

The `entry_points()` function used is actually imported from `backports.entry_points_selectable` (a new dependency, added to `setup.py`), which ensures that calls to `entry_points()` can take filtering arguments even in Python < 3.10, where `importlib.metadata` didn't natively support such usage.

(In practice, for recent Python 3.10+ releases, `backports.entry_points_selectable.entry_points` will be `importlib.metadata.entry_points`.)

The resulting `hotdoc` package passes all unit tests in `hotdoc.tests` on Python 3.12, even in a virtual environment with the `setuptools` package removed:

```console
$ python3 -m unittest
.............................................................................................................
/usr/lib64/python3.12/tempfile.py:936: ResourceWarning:
 Implicitly cleaning up <TemporaryDirectory '/tmp/tmpe0jsh57p'>
  _warnings.warn(warn_message, ResourceWarning)
............
----------------------------------------------------------------------
Ran 121 tests in 3.232s

OK
```

`hotdoc` from the `master` branch, in that same situation, fails several tests outright, and skips over 80% of them:

```console
$ python3 -m unittest
[...]
======================================================================
ERROR: hotdoc.tests.test_hotdoc (unittest.loader._FailedTest.hotdoc.tests.test_hotdoc)
----------------------------------------------------------------------
ImportError: Failed to import test module: hotdoc.tests.test_hotdoc
Traceback (most recent call last):
  File "/usr/lib64/python3.12/unittest/loader.py", line 394, in _find_test_path
    module = self._get_module_from_name(name)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.12/unittest/loader.py", line 337, in _get_module_from_name
    __import__(name)
  File ".../hotdoc/tests/test_hotdoc.py", line 30, in <module>
    from hotdoc.utils.utils import touch
  File ".../hotdoc/utils/utils.py", line 37, in <module>
    import pkg_resources
ModuleNotFoundError: No module named 'pkg_resources'


----------------------------------------------------------------------
Ran 19 tests in 0.002s

FAILED (errors=11)
```

ferdnyc commented on hotdoc/extensions/gi/utils.py:
```diff
@@ -1,10 +1,12 @@
 from backports.entry_points_selectable import entry_points
 
 from hotdoc.core.links import Link
+from hotdoc.utils.loggable import info, debug
```
These are added simply because they're used in the `get_language_classes()` function later on, but would have been undefined if they'd actually been called.

ferdnyc commented on hotdoc/extensions/gi/utils.py:
```diff
@@ -200,16 +202,17 @@ def get_language_classes():
+            group='hotdoc.extensions.gi.languages',
+            name='get_language_classes'):
         try:
+            ep_name = entry_point.name
```
This is here (despite `ep_name` never being used) because the results of `importlib.metadata.entry_points()` are [documented](https://importlib-metadata.readthedocs.io/en/latest/migration.html#pkg-resources-iter-entry-points) as not being validated; invoking any property is the suggested means of triggering validation.

ferdnyc commented on hotdoc/extensions/gi/utils.py:
```diff
@@ -200,16 +202,17 @@ def get_language_classes():
         # pylint: disable=broad-except
         except Exception as exc:
-            info("Failed to load %s" % entry_point.module_name, exc)
+            info("Failed to load %s" % str(entry_point), exc)
```
By the same token, `entry_point.name` (the equivalent of `entry_point.module_name` from `pkg_resources` isn't necessarily going to exist on every result from `entry_points()`, so we can't rely on it for logging. Logging the string value of `entry_point` ensures that it'll be logged even if it isn't a valid `EntryPoint` object.

ferdnyc commented on hotdoc/utils/utils.py:
```diff
@@ -148,22 +149,23 @@ def __get_extra_extension_classes(paths):
-    for dist in distributions:
-        sys.path.append(dist.location)
-        wset.add(dist)
+    sys_paths = set(sys.path)
+    extra_paths = set([Path(x) for x in paths])
+    sys.path.extend(list(extra_paths - sys_paths))
+    del sys_paths
+    del extra_paths
```
Adding `pathlib.Path` objects to `sys.path` [is explicitly supported](https://docs.python.org/3/library/importlib.metadata.html#distribution-discovery) by `importlib.metadata`'s distribution package discovery, "even though such values will be ignored for imports." Since `importlib.metadata` doesn't have the same `WorkingSet` support as `pkg_resources`, that becomes our replacement for the custom `WorkingSet`.

I'm admittedly only 90% sure this actually works like it's supposed to. A real-world test from someone who uses `hotdoc --extra-extension-path` and/or `hotdoc --extra-extension` would not be unwelcome.

ferdnyc replied:
> I'm admittedly only 90% sure this actually works like it's supposed to. A real-world test from someone who uses `hotdoc --extra-extension-path` and/or `hotdoc --extra-extension` would not be unwelcome.

I say that, but running `hotdoc.extensions.gi.utils.get_language_classes()` from the `master` branch in Python 3.11, and from this PR branch in Python 3.12, produces the same results. So, that's a good sign that the basic idea is sound.

It's mostly just the `sys.path` fiddling in `hotdoc.utils.utils.__get_extra_extension_classes()` that I'd like to see validated through real-world testing.

MathieuDuponchelle replied:
This does seem to break `--extra-extension-path`, I put up a reproducer with instructions @ https://github.com/MathieuDuponchelle/my-extension , hopefully this makes it easy to fix :)

ferdnyc replied:
@MathieuDuponchelle Thanks! I'll take a look. I've also since noticed that Meson's docs use some custom extensions, so their code might be useful in testing as well.

ferdnyc replied:
@MathieuDuponchelle 
That repo was a huge help, thanks! I believe I've now tackled this.

ferdnyc commented on hotdoc/utils/utils.py:
```diff
@@ -204,16 +206,17 @@ def get_extension_classes(sort: bool,
-    for entry_point in pkg_resources.iter_entry_points(
+    for entry_point in entry_points(
             group='hotdoc.extensions', name='get_extension_classes'):
-        if entry_point.module_name == 'hotdoc_c_extension.extensions':
-            continue
         try:
+            if entry_point.module == 'hotdoc_c_extension.extensions':
+                continue
+            ep_name = entry_point.name
```
Same validation issue here; since we can't be sure `entry_point` _has_ a `.module` property, we need to check it inside the `try:`.

ferdnyc reviewed (commented):
Some notes.

ferdnyc commented:
For the record, I didn't regenerate the pinned `requirements.txt` with `pip-compile`, because that process seems somewhat broken. Pinning requirements versions, though it sounds nice in theory, often has the unwanted side-effect of locking you in to a particular Python version.

In this case, `pip install -r requirements.txt` **fails** under Python 3.12, because the pinned `lxml` version isn't compatible.

Fortunately, `pip install -e .` uses the unpinned dependency from `setup.py`, so it installs an `lxml` that's compatible with Python 3.12.

MathieuDuponchelle commented:
Thanks A TON for looking at this :)

ferdnyc commented:
I believe this is now fixed. I also added `importlib_metadata` as an explicit dependency in Python < 3.10, even though `backports.entry_points_selectable` will always pull it in if needed.

(I'm now importing it explicitly to access its `distributions()` function, instead of relying solely on the backports `entry_points()`, so I figured it made sense to be explicit about the dependency.)

ferdnyc commented on hotdoc/utils/utils.py:
```diff
@@ -143,39 +149,55 @@ def get_mtime(filename):
+        classes = activation_function()
+    # pylint: disable=broad-except
+    except Exception as exc:
+        if is_installed:
+            info(f"Failed to load {entry_point}", exc)
+        else:
+            warn('extension-import', f"Failed to load {entry_point} {exc}")
```
Oh, I did sneak in a couple of f-strings, which implicitly make hotdoc compatible with Python 3.6+ only.

All earlier Python versions are past end-of-life and no longer supported (heck even Python 3.6 is EOL'd), but I did notice there's still a Python 3.5 check in `setup.py`. If compatibility with older versions is intended, I can take those back out.

ferdnyc replied:
Never mind, there were already f-strings in the file.

MathieuDuponchelle commented:
lgtm now, huge thanks again for performing maintenance on this project :)
